File: ProcessEngine/core/src/test/java/nl/adaptivity/process/engine/TestProcessEngine.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.engine;

import kotlin.jvm.functions.Function2;
import net.devrieze.util.*;
import net.devrieze.util.security.SecureObject;
import net.devrieze.util.security.SimplePrincipal;
import nl.adaptivity.messaging.EndpointDescriptor;
import nl.adaptivity.messaging.EndpointDescriptorImpl;
import nl.adaptivity.process.MemTransactionedHandleMap;
import nl.adaptivity.process.StubTransaction;
import nl.adaptivity.process.engine.ProcessInstance.State;
import nl.adaptivity.process.engine.processModel.IProcessNodeInstance.NodeInstanceState;
import nl.adaptivity.process.engine.processModel.ProcessNodeInstance;
import nl.adaptivity.process.engine.processModel.ProcessNodeInstance.Builder;
import nl.adaptivity.process.processModel.engine.ExecutableProcessNode;
import nl.adaptivity.process.processModel.engine.IProcessModelRef;
import nl.adaptivity.process.processModel.engine.ProcessModelImpl;
import nl.adaptivity.process.processModel.engine.StartNodeImpl;
import nl.adaptivity.util.activation.Sources;
import nl.adaptivity.xml.*;
import org.custommonkey.xmlunit.DetailedDiff;
import org.custommonkey.xmlunit.Diff;
import org.custommonkey.xmlunit.XMLUnit;
import org.jetbrains.annotations.NotNull;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

import javax.xml.bind.JAXB;
import javax.xml.namespace.QName;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.dom.DOMResult;
import javax.xml.transform.dom.DOMSource;

import java.io.*;
import java.net.URI;
import java.util.List;
import java.util.UUID;

import static java.nio.charset.Charset.defaultCharset;
import static nl.adaptivity.process.engine.ProcessInstance.State.FINISHED;
import static nl.adaptivity.process.engine.ProcessInstance.State.STARTED;
import static nl.adaptivity.process.engine.processModel.IProcessNodeInstance.NodeInstanceState.Complete;
import static nl.adaptivity.process.engine.processModel.IProcessNodeInstance.NodeInstanceState.Pending;
import static org.custommonkey.xmlunit.XMLAssert.assertXMLEqual;
import static org.testng.Assert.*;


/**
 * Created by pdvrieze on 18/08/15.
 */
public class TestProcessEngine {

  private static final Function2<SecureObject<ProcessNodeInstance<StubProcessTransaction>>, Long, SecureObject<ProcessNodeInstance<StubProcessTransaction>>> PNI_SET_HANDLE = new Function2<SecureObject<ProcessNodeInstance<StubProcessTransaction>>, Long, SecureObject<ProcessNodeInstance<StubProcessTransaction>>>() {
    @Override
    public SecureObject<ProcessNodeInstance<StubProcessTransaction>> invoke(final SecureObject<ProcessNodeInstance<StubProcessTransaction>> pni, final Long handle) {
      if (pni.withPermission().getHandleValue() == handle) { return pni; }
      final Builder<StubProcessTransaction, ? extends ExecutableProcessNode> builder = pni.withPermission().builder();
      builder.setHandle(Handles.<SecureObject<? extends ProcessNodeInstance<StubProcessTransaction>>>handle(handle));
      return builder.build();
    }
  };
  private static DocumentBuilder _documentBuilder;

  ProcessEngine<StubProcessTransaction> mProcessEngine;
  private StubMessageService mStubMessageService;
  private final EndpointDescriptor mLocalEndpoint = new EndpointDescriptorImpl(QName.valueOf("processEngine"),"processEngine", URI.create("http://localhost/"));
  private ProcessTransactionFactory<StubProcessTransaction> mStubTransactionFactory;
  private SimplePrincipal mPrincipal;

  public TestProcessEngine() {
    mStubMessageService = new StubMessageService(mLocalEndpoint);
    mStubTransactionFactory = new ProcessTransactionFactory<StubProcessTransaction>() {
      @NotNull
      @Override
      public StubProcessTransaction startTransaction(@NotNull final IProcessEngineData<StubProcessTransaction> engineData) {
        return new StubProcessTransaction(engineData);
      }
    };
    mPrincipal = new SimplePrincipal("pdvrieze");
  }

  private InputStream getXml(final String name) {
    try (InputStream in = getClass().getResourceAsStream("/nl/adaptivity/process/engine/test/"+name)) {
      final ByteArrayOutputStream out = new ByteArrayOutputStream();
      if (!InputStreamOutputStream.getInputStreamOutputStream(in, out).get().booleanValue()) {
        return null;
      }
      final byte[] byteArray = out.toByteArray();
      assertTrue(byteArray.length > 0, "Some bytes in the xml files are expected");;
      final ByteArrayInputStream bais = new ByteArrayInputStream(byteArray);
      return bais;
    } catch (Exception e) {
      if (e instanceof  RuntimeException) { throw (RuntimeException) e; }
      throw new RuntimeException(e);
    }
  }

  private XmlReader getStream(final String name) throws XmlException {
    return XmlStreaming.newReader(getXml(name), "UTF-8");
  }

  private ProcessModelImpl getProcessModel(final String name) throws XmlException {
    return ProcessModelImpl.deserialize(getStream(name));
  }

  private Document getDocument(final String name) {
    try (InputStream in = getClass().getResourceAsStream("/nl/adaptivity/process/engine/test/"+name)) {
      return getDocumentBuilder().parse(in);
    } catch (Exception e) {
      if (e instanceof  RuntimeException) { throw (RuntimeException) e; }
      throw new RuntimeException(e);
    }
  }

  private static <V> MutableTransactionedHandleMap<V,StubProcessTransaction> cacheInstances(final MutableTransactionedHandleMap<V,StubProcessTransaction> base, final int count) {
    return new CachingHandleMap<>(base, count);
  }

  private static <V> MutableTransactionedHandleMap<SecureObject<ProcessNodeInstance<StubProcessTransaction>>,StubProcessTransaction> cacheNodes(final MutableTransactionedHandleMap<SecureObject<ProcessNodeInstance<StubProcessTransaction>>,StubProcessTransaction> base, final int count) {
    return new CachingHandleMap<>(base, count, PNI_SET_HANDLE);
  }

  private static <V> IMutableProcessModelMap<StubProcessTransaction> cacheModels(final IMutableProcessModelMap<StubProcessTransaction> base, final int count) {
    return new CachingProcessModelMap<>(base, count);
  }

  @BeforeMethod
  public void beforeTest() {
    mStubMessageService.clear();
//    DelegateProcessEngineData<StubProcessTransaction> engineData =
//            new DelegateProcessEngineData<>(mStubTransactionFactory,
//                                            cache(new MemProcessModelMap(), 1),
//                                            cache(new MemTransactionedHandleMap<>(), 1),
//                                            cache(new MemTransactionedHandleMap<>(), 2));

    mProcessEngine = ProcessEngine.Companion.newTestInstance(
            mStubMessageService,
            mStubTransactionFactory,
            cacheModels(new MemProcessModelMap(), 1),
            cacheInstances(new MemTransactionedHandleMap<SecureObject<ProcessInstance<StubProcessTransaction>>, StubProcessTransaction>(), 1),
            cacheNodes(new MemTransactionedHandleMap<SecureObject<ProcessNodeInstance<StubProcessTransaction>>, StubProcessTransaction>(PNI_SET_HANDLE), 2), true);
  }

  private char[] serializeToXmlCharArray(final Object object) throws XmlException {
    final char[] receivedChars;
    {
      final CharArrayWriter caw = new CharArrayWriter();
      if (object instanceof XmlSerializable) {
        final XmlWriter writer = XmlStreaming.newWriter(caw);
        ((XmlSerializable) object).serialize(writer);
        writer.close();
      } else {
        JAXB.marshal(object, caw);
      }
      receivedChars = caw.toCharArray();
    }
    return receivedChars;
  }

  private static DocumentBuilder getDocumentBuilder() {
    if (_documentBuilder==null) {
      final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
      try {
        dbf.setNamespaceAware(true);
        dbf.setIgnoringElementContentWhitespace(false);
        dbf.setCoalescing(false);
        _documentBuilder = dbf.newDocumentBuilder();
      } catch (ParserConfigurationException e) {
        throw new RuntimeException(e);
      }
    }
    return _documentBuilder;
  }

  @Test
  public void testExecuteSingleActivity() throws Exception {
    final ProcessModelImpl       model       = getProcessModel("testModel1.xml");
    final StubProcessTransaction transaction = mProcessEngine.startTransaction();
    final IProcessModelRef       modelHandle = mProcessEngine.addProcessModel(transaction, model, mPrincipal);

    final HProcessInstance instanceHandle = mProcessEngine.startProcess(transaction, mPrincipal, modelHandle, "testInstance1", UUID.randomUUID(), null);

    assertEquals(mStubMessageService.getMMessages().size(), 1);
    assertEquals(mStubMessageService.getMessageNode(0).getHandleValue(), 1L);

    final InputStream expected = getXml("testModel1_task1.xml");

    final char[] receivedChars = serializeToXmlCharArray(mStubMessageService.getMMessages().get(0));

    XMLUnit.setIgnoreWhitespace(true);
    try {
      assertXMLEqual(new InputStreamReader(expected), new CharArrayReader(receivedChars));
    } catch (AssertionError e) {
      e.printStackTrace();
      try {
        assertEquals(new String(receivedChars), Streams.toString(getXml("testModel1_task1.xml"), defaultCharset()));
      } catch (Exception f) {
        f.initCause(e);
        throw f;
      }

    }

    ProcessInstance<StubProcessTransaction> processInstance = mProcessEngine.getProcessInstance(transaction, instanceHandle, mPrincipal);
    assertEquals(processInstance.getState(), State.STARTED);

    assertEquals(processInstance.getActive().size(), 1);
    assertEquals(processInstance.getFinished().size(), 1);
    final Handle<? extends SecureObject<ProcessNodeInstance<StubProcessTransaction>>> hfinished = processInstance.getFinished().iterator().next();
    final ProcessNodeInstance<StubProcessTransaction>                                 finished  = mProcessEngine.getNodeInstance(transaction, hfinished, mPrincipal);
    assertTrue(finished.getNode() instanceof StartNodeImpl);;
    assertEquals(finished.getNode().getId(), "start");

    assertEquals(processInstance.getResults().size(), 0);

    final ProcessNodeInstance taskNode = mProcessEngine.getNodeInstance(transaction, mStubMessageService.getMessageNode(0), mPrincipal);
    assertEquals(taskNode.getState(), NodeInstanceState.Pending); // Our messenger does not do delivery notification

    assertEquals(mProcessEngine.finishTask(transaction, taskNode.getHandle(), null, mPrincipal).getState(), NodeInstanceState.Complete);
    processInstance = mProcessEngine.getProcessInstance(transaction, instanceHandle, mPrincipal);
    assertEquals(processInstance.getActive().size(), 0);
    assertEquals(processInstance.getFinished().size(), 2);
    assertEquals(processInstance.getResults().size(), 1);

    assertEquals(processInstance.getState(), State.FINISHED);
  }

  @Test
  public void testGetDataFromTask() throws Exception {
    final ProcessModelImpl       model       = getProcessModel("testModel2.xml");
    final StubProcessTransaction transaction = mProcessEngine.startTransaction();
    final IProcessModelRef       modelHandle = mProcessEngine.addProcessModel(transaction, model, mPrincipal);

    final HProcessInstance instanceHandle = mProcessEngine.startProcess(transaction, mPrincipal, modelHandle, "testInstance1", UUID.randomUUID(), null);

    assertEquals(mStubMessageService.getMMessages().size(), 1);

    XMLUnit.setIgnoreWhitespace(true);
    assertXMLEqual(new InputStreamReader(getXml("testModel2_task1.xml")), new CharArrayReader(serializeToXmlCharArray(mStubMessageService
                                                                                                                              .getMMessages()
                                                                                                                              .get(0))));
    ProcessNodeInstance<StubProcessTransaction> ac1 = mProcessEngine.getNodeInstance(transaction, mStubMessageService.getMessageNode(0), mPrincipal);// This should be 0 as it's the first activity


    mStubMessageService.clear(); // (Process the message)
    assertEquals(ac1.getResults().size(), 0);
    ac1 = mProcessEngine.finishTask(transaction, ac1.getHandle(), getDocument("testModel2_response1.xml"), mPrincipal);
    assertEquals(ac1.getState(), NodeInstanceState.Complete);
    ac1 = mProcessEngine.getNodeInstance(transaction, ac1.getHandle(), mPrincipal);
    assertEquals(ac1.getResults().size(), 2);
    final ProcessData result1 = ac1.getResults().get(0);
    final ProcessData result2 = ac1.getResults().get(1);
    assertEquals(result1.getName(), "name");
    assertEquals(result1.getContent().getContentString(), "Paul");
    assertEquals(result2.getName(), "user");
    assertXMLEqual("<user><fullname>Paul</fullname></user>", result2.getContent().getContentString());

    assertEquals(mStubMessageService.getMMessages().size(), 1);
    assertEquals(mStubMessageService.getMessageNode(0).getHandleValue(), 2L); //We should have a new message with the new task (with the data)
    final ProcessNodeInstance ac2 =mProcessEngine.getNodeInstance(transaction, mStubMessageService.getMessageNode(0), mPrincipal);

    final List<ProcessData> ac2Defines = ac2.getDefines(transaction);
    assertEquals(ac2Defines.size(), 1);


    final ProcessData define = ac2Defines.get(0);
    assertEquals(define.getName(), "mylabel");
    assertEquals(define.getContent().getContentString(), "Hi Paul. Welcome!");

  }

  private static void assertXMLSimilar(final Document expected, final Document actual) {
    final Diff         diff         = XMLUnit.compareXML(expected, actual);
    final DetailedDiff detailedDiff = new DetailedDiff(diff);
    if(! detailedDiff.similar()) {
      fail(detailedDiff.toString());
    }
  }

  private static Document toDocument(final Node node) throws TransformerException {
    final Document result = getDocumentBuilder().newDocument();
    Sources.writeToResult(new DOMSource(node), new DOMResult(result));
    return result;
  }


}

File: ProcessEngine/core/src/test/kotlin/nl/adaptivity/process/engine/TestProcessEngine.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the 
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.engine

import net.devrieze.util.*
import net.devrieze.util.security.SecureObject
import net.devrieze.util.security.SimplePrincipal
import nl.adaptivity.dropStack
import nl.adaptivity.messaging.EndpointDescriptorImpl
import nl.adaptivity.process.MemTransactionedHandleMap
import nl.adaptivity.process.engine.ProcessInstance.State
import nl.adaptivity.process.engine.processModel.IProcessNodeInstance
import nl.adaptivity.process.engine.processModel.NodeInstanceState
import nl.adaptivity.process.engine.processModel.ProcessNodeInstance
import nl.adaptivity.process.processModel.XmlMessage
import nl.adaptivity.process.processModel.engine.ExecutableProcessModel
import nl.adaptivity.process.processModel.engine.ExecutableStartNode
import nl.adaptivity.util.activation.Sources
import nl.adaptivity.xml.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.xml.sax.InputSource
import org.xmlunit.XMLUnitException
import org.xmlunit.builder.DiffBuilder
import org.xmlunit.diff.DefaultComparisonFormatter
import java.io.*
import java.net.URI
import java.util.UUID
import javax.xml.bind.JAXB
import javax.xml.namespace.QName
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerException
import javax.xml.transform.dom.DOMResult
import javax.xml.transform.dom.DOMSource
import kotlin.collections.ArrayList


/**
 * Created by pdvrieze on 18/08/15.
 */
class TestProcessEngine {

    internal lateinit var mProcessEngine: ProcessEngine<StubProcessTransaction>
    private val localEndpoint = EndpointDescriptorImpl(QName.valueOf("processEngine"),
                                                       "processEngine",
                                                       URI.create("http://localhost/"))
    private val stubMessageService: StubMessageService = StubMessageService(localEndpoint)
    private val stubTransactionFactory = object : ProcessTransactionFactory<StubProcessTransaction> {
        override fun startTransaction(engineData: IProcessEngineData<StubProcessTransaction>): StubProcessTransaction {
            return StubProcessTransaction(engineData)
        }
    }
    private val principal = SimplePrincipal("pdvrieze")

    private fun getXml(name: String): ByteArray? {
        javaClass.getResourceAsStream("/nl/adaptivity/process/engine/test/" + name).use { reader ->
            val out = ByteArrayOutputStream()
            if (!InputStreamOutputStream.getInputStreamOutputStream(reader, out).get()) {
                return null
            }
            val byteArray = out.toByteArray()
            assertTrue(byteArray.isNotEmpty(), "Some bytes in the xml files are expected")
            return byteArray
        }
    }

    @Throws(XmlException::class)
    private fun getStream(name: String): XmlReader {
        return XmlStreaming.newReader(ByteArrayInputStream(getXml(name)!!), "UTF-8")
    }

    @Throws(XmlException::class)
    private fun getProcessModel(name: String): ExecutableProcessModel {
        return ExecutableProcessModel.deserialize(getStream(name))
    }

    private fun getDocument(name: String): Document {
        try {
            javaClass.getResourceAsStream("/nl/adaptivity/process/engine/test/" + name).use { reader -> return documentBuilder.parse(reader) }
        } catch (e: Exception) {
            if (e is RuntimeException) {
                throw e
            }
            throw RuntimeException(e)
        }

    }

    @Throws(XmlException::class)
    private fun serializeToXmlCharArray(obj: Any): CharArray {
        return CharArrayWriter().apply {
            val caw = this
            if (obj is XmlSerializable) {
                XmlStreaming.newWriter(this).use { writer ->
                    obj.serialize(writer)
                }
            } else {
                JAXB.marshal(obj, caw)
            }
        }.toCharArray()
    }

    private inline fun <R> testProcess(model: ExecutableProcessModel, payload: Node? = null, body: (ProcessTransaction, ExecutableProcessModel, HProcessInstance) -> R):R {
        mProcessEngine.startTransaction().use { transaction ->

            val modelHandle = mProcessEngine.addProcessModel(transaction, model, principal)
            val instanceHandle = mProcessEngine.startProcess(transaction, principal, modelHandle, "testInstance", UUID.randomUUID(), payload)

            return body(transaction, transaction.readableEngineData.processModel(modelHandle).mustExist(modelHandle).withPermission(), instanceHandle)
        }
    }

    private fun ProcessTransaction.getInstance(instanceHandle: HProcessInstance):ProcessInstance {
        return readableEngineData.instance(instanceHandle).withPermission()
    }

    private fun StubMessageService.messageNode(transaction: ProcessTransaction, index:Int): ProcessNodeInstance<*> {
        return transaction.readableEngineData.nodeInstance(this._messages[index].source).withPermission()
    }

    private fun ProcessInstance.child(transaction: ProcessTransaction, name: String) : ProcessNodeInstance<*> {
        return getChild(name, 1)?.withPermission() ?: throw AssertionError("No node instance for node id ${name} found")
    }

    private val ProcessInstance.sortedFinished
        get() = finished.sortedBy { it.handleValue }

    private val ProcessInstance.sortedActive
        get() = active.sortedBy { it.handleValue }

    private val ProcessInstance.sortedCompleted
        get() = completedEndnodes.sortedBy { it.handleValue }

    private fun ProcessInstance.assertFinishedHandles(vararg handles: ComparableHandle<SecureObject<ProcessNodeInstance<*>>>) = apply {
        assertEquals(handles.sorted(), ArrayList(sortedFinished))
    }

    private fun ProcessInstance.assertFinished(vararg handles: IProcessNodeInstance) = apply {
        val actual = ArrayList(sortedFinished)
        val expected = handles.asSequence().map { it.handle() }.sortedBy { it.handleValue }.toList()
        try {
            assertEquals(expected, actual)
        } catch (e: AssertionError) {
            throw AssertionError("Expected finished list $expected, but found $actual").initCause(e)
        }
    }

    private fun ProcessInstance.assertActiveHandles(vararg handles: ComparableHandle<SecureObject<ProcessNodeInstance<*>>>) = apply {
        assertEquals(handles.sorted(), ArrayList(sortedActive))
    }

    private fun ProcessInstance.assertActive(vararg handles: IProcessNodeInstance) = apply {
        val actual = ArrayList(sortedActive)
        val expected = handles.asSequence().map { it.handle() }.sortedBy { it.handleValue }.toList()
        try {
            assertEquals(expected, actual)
        } catch (e:AssertionError) {
            throw AssertionError("Expected active list $expected, but found $actual").initCause(e)
        }
    }

    private fun ProcessInstance.assertCompletedHandles(vararg handles: ComparableHandle<SecureObject<ProcessNodeInstance<*>>>) = apply {
        assertEquals(handles.sorted(), ArrayList(sortedCompleted))
    }

    private fun ProcessInstance.assertCompleted(vararg nodes: IProcessNodeInstance): ProcessInstance {
        val actual = ArrayList(sortedCompleted)
        val expected = nodes.asSequence().map { it.handle() }.sortedBy { it.handleValue }.toList()
        assertEquals(expected, actual)
        return this
    }

    private fun IProcessNodeInstance.assertStarted() = apply {
        assertEquals(NodeInstanceState.Started, this.state)
    }

    private fun ProcessInstance.assertIsStarted() = apply {
        assertEquals(State.STARTED, this.state)
    }

    private fun ProcessInstance.assertIsFinished() = apply {
        assertEquals(State.FINISHED, this.state)
    }

    private fun IProcessNodeInstance.assertSent() = apply {
        assertState(NodeInstanceState.Sent)
    }

    private fun IProcessNodeInstance.assertPending() = apply {
        assertState(NodeInstanceState.Pending)
    }

    private fun IProcessNodeInstance.assertAcknowledged() = apply {
        assertState(NodeInstanceState.Acknowledged)
    }

    private fun IProcessNodeInstance.assertComplete() = apply {
        assertState(NodeInstanceState.Complete)
    }

    private fun IProcessNodeInstance.assertState(state: NodeInstanceState) {
        try {
            assertEquals(state, this.state,
                         "Node ${this.node.id}(${this.handle()}) should be in the ${state.name} state")
        } catch (e: AssertionError) {
            throw dropStack(e, 2)
        }
    }

    @BeforeEach
    fun beforeTest() {
        stubMessageService.clear()
        //    DelegateProcessEngineData<StubProcessTransaction> engineData =
        //            new DelegateProcessEngineData<>(mStubTransactionFactory,
        //                                            cache(new MemProcessModelMap(), 1),
        //                                            cache(new MemTransactionedHandleMap<>(), 1),
        //                                            cache(new MemTransactionedHandleMap<>(), 2));

        mProcessEngine = ProcessEngine.newTestInstance(
            stubMessageService,
            stubTransactionFactory,
            cacheModels<Any>(MemProcessModelMap(), 3),
            cacheInstances(MemTransactionedHandleMap<SecureObject<ProcessInstance>, StubProcessTransaction>(), 1),
            cacheNodes<Any>(MemTransactionedHandleMap<SecureObject<ProcessNodeInstance<*>>, StubProcessTransaction>(PNI_SET_HANDLE), 2), true)
    }

    fun assertEqualsXml(expected: ByteArray, actual: CharArray?) {
        if (actual==null) org.junit.jupiter.api.fail("No actual result")
        val expected = String(expected)
        val actual = String(actual)

        val expectedDoc = DocumentBuilderFactory
            .newInstance().apply { isNamespaceAware=true }
            .newDocumentBuilder()
            .parse(InputSource(StringReader(expected)))
        assertNotNull(expectedDoc)
        val diff = try {
            DiffBuilder.compare(expectedDoc)
                .withTest(actual)
                .ignoreWhitespace()
                .ignoreComments()
                .checkForSimilar()
                .build()
        } catch (e: XMLUnitException) {
            assertEquals(expected, actual, "Error comparing: ${e.message}")
            throw e
        }

        if (diff.hasDifferences()) {
            assertEquals(expected, actual,
                         diff.toString(DefaultComparisonFormatter()))
        }
    }

    fun assertEqualsXml(expected: String?, actual: String?) {
        val diff = DiffBuilder.compare(expected)
            .withTest(actual)
            .ignoreWhitespace()
            .ignoreComments()
            .checkForSimilar()
            .build()

        if (diff.hasDifferences()) {
            assertEquals(expected, actual, diff.toString(DefaultComparisonFormatter()))
        }
    }

    @Test
    @Throws(Exception::class)
    fun testExecuteSingleActivity() {
        val model = getProcessModel("testModel1.xml")
        val transaction = mProcessEngine.startTransaction()
        val engineData = transaction.writableEngineData
        val modelHandle = mProcessEngine.addProcessModel(transaction, model, principal)

        val instanceHandle = mProcessEngine.startProcess(transaction, principal, modelHandle, "testInstance1", UUID.randomUUID(), null)

        assertEquals(1, stubMessageService._messages.size)
        assertEquals(1L, stubMessageService.getMessageNode(0).handleValue)

        val expected = getXml("testModel1_task1.xml")!!

        val receivedChars = serializeToXmlCharArray(stubMessageService._messages[0].base)

        assertEqualsXml(expected, receivedChars)

        run {
            val processInstance = transaction.getInstance(instanceHandle).assertIsStarted()

            assertEquals(1, processInstance.active.size)

            processInstance.child(transaction, "start").assertComplete().let { start ->
                processInstance.assertFinished(start)
                assertTrue(start.node is ExecutableStartNode)
            }

            processInstance.assertCompleted() // no completions

            val taskNode = stubMessageService.messageNode(transaction, 0)
            taskNode.assertAcknowledged()
            processInstance.assertActive(taskNode)
            processInstance.update(engineData) {
                updateChild(taskNode) {
                    finishTask(engineData)
                    assertComplete()
                }
            }
        }

        run {
            val processInstance = transaction.getInstance(instanceHandle).assertIsFinished()
            val start = processInstance.child(transaction, "start")
            val ac = processInstance.child(transaction, "ac2")
            val end = processInstance.child(transaction, "end")
            processInstance.assertActive()
            processInstance.assertFinished(start, ac)
            processInstance.assertCompleted(end)

        }
    }

    @Test
    fun testConditionFalse() {
        val model = ExecutableProcessModel.build {
            owner = principal
            val start = startNode { id="start" }
            val ac = activity { id="ac"; predecessor=start.identifier; condition="false()" }
            val end = endNode { id="end"; predecessor=ac }
        }
        testProcess(model) { transaction, model, instanceHandle ->
            transaction.readableEngineData.instance(instanceHandle).withPermission().let { instance ->
                val start = instance.child(transaction, "start")
                val ac = instance.child(transaction, "ac").apply { assertState(NodeInstanceState.Skipped) }
                val end = instance.child(transaction, "end").apply { assertState(NodeInstanceState.Skipped) }
                instance.assertFinished(start, ac)
                assertEquals(State.SKIPPED, instance.state)
            }
        }
    }

    @Test
    fun testConditionTrue() {
        val model = ExecutableProcessModel.build {
            owner = principal
            val start = startNode { id="start" }
            val ac = activity { id="ac"; predecessor=start.identifier; condition="true()" }
            val end = endNode { id="end"; predecessor=ac }
        }
        testProcess(model) { transaction, model, instanceHandle ->
            transaction.readableEngineData.instance(instanceHandle).withPermission().let { instance ->
                val start = instance.child(transaction, "start")
                val ac = instance.child(transaction, "ac").assertAcknowledged()
//        val end = instance.child(transaction, "end").apply { assertState(NodeInstanceState.Skipped) }
                instance.assertFinished(start)
                assertEquals(State.STARTED, instance.state)
            }
        }
    }

    @Test
    fun testSplitJoin1() {
        testProcess(simpleSplitModel) { transaction, model, instanceHandle ->
            val engineData = transaction.writableEngineData
            run {
                run {
                    val instance = transaction.readableEngineData.instance(instanceHandle).withPermission()
                    val start = instance.child(transaction, "start")
                    val split = instance.child(transaction, "split1").assertStarted()
                    val ac1 = instance.child(transaction, "ac1").assertAcknowledged()
                    val ac2 = instance.child(transaction, "ac2").assertAcknowledged()

                    instance.assertFinished(start)
                    instance.assertActive(split, ac1, ac2)

                    run {
                        val messageSources = stubMessageService._messages
                            .map { transaction.readableEngineData.nodeInstance(it.source).withPermission() }
                            .sortedBy { it.node.id }
                        assertEquals(listOf(ac1, ac2), messageSources)
                        stubMessageService._messages.forEach { msg ->
                            msg.source
                        }
                    }

                    instance.update(engineData) {
                        updateChild(ac1) {
                            finishTask(engineData)
                            assertComplete()
                        }
                    }

                }
                run {
                    val instance = transaction.readableEngineData.instance(instanceHandle).withPermission()
                    val start = instance.child(transaction, "start")
                    val split = instance.child(transaction, "split1").assertStarted()
                    val ac1 = instance.child(transaction, "ac1").assertComplete()
                    val ac2 = instance.child(transaction, "ac2").assertAcknowledged()
                    val join = instance.child(transaction, "join1").assertPending()
                    instance.assertFinished(ac1, start)
                    instance.assertActive(ac2, split, join)
                    // check join is in the pending set

                    instance.update(transaction.writableEngineData) {
                        updateChild(ac2) {
                            startTask(transaction.writableEngineData)
                        }
                    }
                }
                run {
                    val instance = transaction.readableEngineData.instance(instanceHandle).withPermission()
                    val start = instance.child(transaction, "start")
                    val split = instance.child(transaction, "split1").assertStarted()
                    val ac1 = instance.child(transaction, "ac1").assertComplete()
                    val ac2 = instance.child(transaction, "ac2").assertStarted()
                    val join = instance.child(transaction, "join1").assertPending()
                    instance.assertFinished(ac1, start)
                    instance.assertActive(ac2, split, join)

                    instance.update(engineData) {
                        updateChild(ac2) {
                            finishTask(engineData)
                            assertComplete()
                        }
                    }
                }

                run {
                    val instance = transaction.readableEngineData.instance(instanceHandle).withPermission()
                    val start = instance.child(transaction, "start")
                    val split = instance.child(transaction, "split1").assertComplete()
                    val ac1 = instance.child(transaction, "ac1").assertComplete()
                    val ac2 = instance.child(transaction, "ac2").assertComplete()
                    val join = instance.child(transaction, "join1").assertComplete()
                    val end = instance.child(transaction, "end").assertComplete()

                    assertEquals(0, instance.active.size)
                    instance.assertFinished(start, split, ac1, ac2, join)
                    instance.assertCompleted(end)
                }

            }
            run {
                val instance = transaction.readableEngineData.instance(instanceHandle).withPermission()
                assertEquals(State.FINISHED, instance.state)
            }
        }

    }

    @Test
    @Throws(Exception::class)
    fun testGetDataFromTask() {
        val model = getProcessModel("testModel2.xml")
        val transaction = mProcessEngine.startTransaction()
        val engineData = transaction.writableEngineData
        val modelHandle = mProcessEngine.addProcessModel(transaction, model, principal)

        val instanceHandle = mProcessEngine.startProcess(transaction, principal, modelHandle, "testInstance1", UUID.randomUUID(), null)

        assertEquals(1, stubMessageService._messages.size)

        assertEqualsXml(getXml("testModel2_task1.xml")!!, serializeToXmlCharArray(stubMessageService
                                                                                      ._messages[0].base))

        var ac1: ProcessNodeInstance<*> = mProcessEngine.getNodeInstance(transaction, stubMessageService.getMessageNode(0), principal) ?: throw AssertionError("Message node not found")// This should be 0 as it's the first activity

        ac1.node.results.let { r ->
            assertEquals("", r[0].contentString)
            assertEquals("name", r[0].getName())
            assertEquals("user", r[1].getName())
            assertEquals("/umh:result/umh:value[@name='user']/text()", r[0].getPath())
            assertEquals(null, r[1].getPath())

            assertEquals(listOf(XmlEvent.NamespaceImpl("umh", "http://adaptivity.nl/userMessageHandler")),
                         r[0].namespaces.sortedBy { it.prefix })
            assertEquals(listOf(XmlEvent.NamespaceImpl("", "http://adaptivity.nl/ProcessEngine/"),
                                XmlEvent.NamespaceImpl("umh", "http://adaptivity.nl/userMessageHandler")),
                         r[1].namespaces.sortedBy { it.prefix })

            val result2ExpectedContent = """|
                |      <user xmlns="" xmlns:jbi="http://adaptivity.nl/ProcessEngine/activity">
                |        <fullname>
                |          <jbi:value xpath="/umh:result/umh:value[@name='user']/text()"/>
                |        </fullname>
                |      </user>
                |    """.trimMargin("|")
            assertEquals(result2ExpectedContent, r[1].contentString)
        }

        stubMessageService.clear() // (Process the message)
        assertEquals(0, ac1.results.size)
        ac1 = mProcessEngine.finishTask(transaction, ac1.getHandle(), getDocument("testModel2_response1.xml"), principal)
        assertEquals(NodeInstanceState.Complete, ac1.state)
        ac1 = mProcessEngine.getNodeInstance(transaction, ac1.getHandle(), principal) ?: throw AssertionError("Node ${ac1.getHandle()} not found")
        assertEquals(2, ac1.results.size)
        val result1 = ac1.results[0]
        val result2 = ac1.results[1]
        assertEquals("name", result1.name)
        assertEquals("Paul", result1.content.contentString)
        assertEquals("user", result2.name)

        result2.content.contentString.let { actual ->
            val expected = "<user xmlns=''><fullname>Paul</fullname></user>"
            assertEqualsXml(expected, actual)
        }
        assertEquals(1, stubMessageService._messages.size)
        assertEquals(2L, stubMessageService.getMessageNode(0).handleValue) //We should have a new message with the new task (with the data)
        val ac2 = mProcessEngine.getNodeInstance(transaction, stubMessageService.getMessageNode(0), principal)

        val ac2Defines = ac2!!.getDefines(engineData)
        assertEquals(1, ac2Defines.size)


        val define = ac2Defines[0]
        assertEquals("mylabel", define.name)
        assertEquals("Hi Paul. Welcome!", define.content.contentString)

    }

    private val simpleSplitModel: ExecutableProcessModel get() {
        return ExecutableProcessModel.build {
            owner = principal
            val start = startNode {
                id = "start"
            }
            val split1 = split {
                predecessor = start.identifier
                id = "split1"
                min = 2
                max = 2
            }
            val ac1 = activity {
                predecessor = split1.identifier
                id = "ac1"
                message = XmlMessage()
                result {
                    name = "ac1result"
                    content = "ac1content".toCharArray()
                }
            }
            val ac2 = activity {
                predecessor = split1.identifier
                id = "ac2"
                message = XmlMessage()
                result {
                    name = "ac2result"
                    content = "ac2content".toCharArray()
                }
            }
            val join = join {
                predecessors(ac1, ac2)
                id = "join1"
                min = 2
                max = 2
            }
            endNode {
                id = "end"
                predecessor = join
            }
        }
    }

    companion object {

        internal val PNI_SET_HANDLE = fun(transaction: StubProcessTransaction, pni: SecureObject<ProcessNodeInstance<*>>, handle: Handle<SecureObject<ProcessNodeInstance<*>>>): SecureObject<ProcessNodeInstance<*>>? {
            if (pni.withPermission().getHandle() == handle) {
                return pni
            }
            val piBuilder = transaction.readableEngineData.instance(pni.withPermission().hProcessInstance).withPermission().builder()
            return pni.withPermission().builder(piBuilder).also { it.handle = Handles.handle(handle) }.build()
        }

        internal fun <V:Any> cacheInstances(base: MutableTransactionedHandleMap<V, StubProcessTransaction>, count: Int): MutableTransactionedHandleMap<V, StubProcessTransaction> {
            return CachingHandleMap<V, StubProcessTransaction>(base, count)
        }

        internal fun <V> cacheNodes(base: MutableTransactionedHandleMap<SecureObject<ProcessNodeInstance<*>>, StubProcessTransaction>, count: Int): MutableTransactionedHandleMap<SecureObject<ProcessNodeInstance<*>>, StubProcessTransaction> {
            return CachingHandleMap(base, count, PNI_SET_HANDLE)
        }

        internal fun <V> cacheModels(base: IMutableProcessModelMap<StubProcessTransaction>, count: Int): IMutableProcessModelMap<StubProcessTransaction> {
            return CachingProcessModelMap(base, count)
        }

        private val documentBuilder: DocumentBuilder by lazy {
            DocumentBuilderFactory.newInstance().apply {
                isNamespaceAware = true
                isIgnoringElementContentWhitespace = false
                isCoalescing = false

            }.newDocumentBuilder()
        }

        @Throws(TransformerException::class)
        private fun toDocument(node: Node): Document {
            val result = documentBuilder.newDocument()
            Sources.writeToResult(DOMSource(node), DOMResult(result))
            return result
        }
    }


}
